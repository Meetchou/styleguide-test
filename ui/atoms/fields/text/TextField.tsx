import React, { FC, ChangeEvent, MouseEvent } from 'react';
import cls from 'classnames';

import { IconType, SizeType, KeyOfSizeType } from '@ui/types';
import { getIcon } from '@ui/atoms/icons/icon/utils';
import InputContainer from '@ui/atoms/containers/input/InputContainer';
import { getFieldClasses } from '@ui/atoms/fields/utils';
import Skeleton from '@ui/atoms/loaders/skeleton/Skeleton';

type IconProps = {
  icon?: IconType;
  onIconClick?: (event: MouseEvent) => void;
};

const sizes: KeyOfSizeType = {
  small: 'p-2',
  base: 'p-3',
  large: 'p-4'
};

export type TextFieldProps = React.HTMLAttributes<HTMLInputElement> & IconProps & {
  id?: string;
  name: string;
  type?: string;
  className?: string;
  label?: string;
  value?: string;
  error?: string;
  placeholder?: string;
  readOnly?: boolean;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
  loading?: boolean;
  borderless?: boolean;
  size?: SizeType;
};

export const TextFieldSkeleton: FC = () => <Skeleton size='large' field />;

const Icon:FC<IconProps> = ({ icon, onIconClick }) => {
  const Icon: React.ComponentType | undefined = getIcon(icon);
  if (!Icon) {
    return null;
  }

  const className = 'absolute inset-y-0 right-0 pr-3 flex items-center text-sm leading-5 text-gray-400';
  if (!onIconClick) {
    return (
      <span className={className}>
        <Icon />
      </span>
    );
  }

  return (
    <span
      className={cls(className, 'hover:text-primary-500 transition-colors cursor-pointer')}
      onClick={onIconClick}
      aria-hidden='true'
    >
      <Icon />
    </span>
  );
};

const TextField:React.ForwardRefRenderFunction<HTMLInputElement, TextFieldProps> = (props, ref) => {
  const {
    className,
    id,
    name,
    type = 'text',
    value,
    placeholder,
    label,
    icon,
    onChange,
    onIconClick,
    error,
    readOnly,
    loading,
    borderless = false,
    size = 'base',
    ...rest
  } = props;

  const inputCls = React.useMemo(
    () => cls(
      getFieldClasses({ active: !readOnly, borderless }),
      {
        'text-danger-500 border-danger-500': !!error && !readOnly,
        'pr-7': !!icon,
        'text-gray-300': readOnly
      }
    ),
    [readOnly, icon, error, borderless]
  );

  return (
    <InputContainer
      htmlFor={id || name}
      error={error}
      label={label}
      loading={loading}
      data-testid={`textfield-${type}-${name}`}
      className={className}
      skeleton={TextFieldSkeleton}
    >
      <input
        {...rest}
        id={id || name}
        ref={ref}
        type={type}
        name={name}
        readOnly={readOnly}
        className={cls(inputCls, sizes[size])}
        value={value}
        placeholder={placeholder}
        onChange={onChange}
      />
      <Icon onIconClick={onIconClick} icon={icon} />
    </InputContainer>
  );
};

export default React.forwardRef(TextField);
