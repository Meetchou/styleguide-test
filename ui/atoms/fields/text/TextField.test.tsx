import React from 'react';

import render from '@ui/testing/render';

import TextField from './TextField';

describe('atoms/fields/text/TextField', () => {
  const props = {
    label: 'label',
    placeholder: 'placeholder',
    name: 'name',
    value: 'value',
    icon: 'eye',
    type: 'text'
  };

  describe('renders correctly', () => {
    it('field', () => {
      const tree = render(<TextField {...props} />);

      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('skeleton', () => {
      const tree = render(<TextField {...props} loading />);

      expect(tree.toJSON()).toMatchSnapshot();
      expect(
        tree.root.findAllByType('input')
      ).toHaveLength(0);
    });
  });

  it('should handle icon click', () => {
    const onIconClick = jest.fn();
    const tree = render(
      <TextField
        {...props}
        onIconClick={onIconClick}
      />
    );

    tree.root.findByType('span').props.onClick();

    expect(onIconClick).toHaveBeenCalled();
  });
});
