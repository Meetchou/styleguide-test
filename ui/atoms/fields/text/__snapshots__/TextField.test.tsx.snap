// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/fields/text/TextField renders correctly field 1`] = `
<div
  data-testid="textfield-text-name"
>
  <label
    className="block text-xs font-semibold text-gray-900 mb-2"
    htmlFor="name"
  >
    label
  </label>
  <div
    className="relative"
  >
    <input
      className="w-full bg-white text-xs font-medium placeholder-gray-400 outline-none rounded-lg border border-gray-100 focus:border-primary-500 active:border-primary-500 pr-7 p-3"
      id="name"
      name="name"
      placeholder="placeholder"
      type="text"
      value="value"
    />
    <span
      className="absolute inset-y-0 right-0 pr-3 flex items-center text-sm leading-5 text-gray-400"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z"
        />
        <circle
          cx="12"
          cy="12"
          r="3"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`atoms/fields/text/TextField renders correctly skeleton 1`] = `
<div
  data-testid="textfield-text-name"
>
  <div
    className="animate-pulse bg-gray-100 rounded h-3 mt-1 mb-2 w-1/2 sm:w-1/4"
  />
  <div
    className="animate-pulse bg-gray-100 rounded-lg h-10"
  />
</div>
`;
